<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>LocARNA-2.0.0RC10: exparna_p</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">LocARNA-2.0.0RC10
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="PageDoc"><div class="header">
  <div class="headertitle">
<div class="title">exparna_p </div>  </div>
</div><!--header-->
<div class="contents">
<div class="toc"><h3>Table of Contents</h3>
<ul><li class="level1"><a href="#autotoc_md107">exparna_p</a></li>
<li class="level1"><a href="#autotoc_md108">NAME</a></li>
<li class="level1"><a href="#autotoc_md109">DESCRIPTION</a><ul><li class="level2"><a href="#autotoc_md110">Input</a></li>
<li class="level2"><a href="#autotoc_md111">Output</a></li>
</ul>
</li>
<li class="level1"><a href="#autotoc_md112">OPTIONS</a><ul><li class="level2"><a href="#autotoc_md113">Scoring parameters:</a></li>
<li class="level2"><a href="#autotoc_md114">Heuristics for speed accuracy trade off:</a></li>
<li class="level2"><a href="#autotoc_md115">Output:</a></li>
<li class="level2"><a href="#autotoc_md116">Suboptimal traceback:</a></li>
<li class="level2"><a href="#autotoc_md117">Constraints:</a></li>
<li class="level2"><a href="#autotoc_md118">Miscellaneous:</a></li>
<li class="level2"><a href="#autotoc_md119">Input files:</a></li>
</ul>
</li>
<li class="level1"><a href="#autotoc_md120">AVAILABILITY</a></li>
<li class="level1"><a href="#autotoc_md121">REFERENCES</a></li>
<li class="level1"><a href="#autotoc_md122">EXAMPLES</a></li>
<li class="level1"><a href="#autotoc_md123">AUTHOR</a></li>
<li class="level1"><a href="#autotoc_md124">REPORTING BUGS</a></li>
<li class="level1"><a href="#autotoc_md125">COPYRIGHT</a></li>
<li class="level1"><a href="#autotoc_md126">SEE ALSO</a></li>
</ul>
</div>
<div class="textblock"><h1><a class="anchor" id="autotoc_md107"></a>
exparna_p</h1>
<h1><a class="anchor" id="autotoc_md108"></a>
NAME</h1>
<p>exparna_p - manual page for exparna_p (LocARNA 2.0.0RC10)</p>
<h1><a class="anchor" id="autotoc_md109"></a>
DESCRIPTION</h1>
<p>exparna_p: fast exact structure local matching of RNAs.</p>
<p>USAGE: exparna_p [options] &lt;Input 1&gt; &lt;Input 2&gt;</p>
<p>exparna_p performs pairwise matching and folding, i.e. it determines exactly matching local favourable sub-structures that can be simulataneously formed in the input RNAs.</p>
<h2><a class="anchor" id="autotoc_md110"></a>
Input</h2>
<p>Input consists of two sequences or alignments, which are specified in fasta, clustal, stockholm, or LocARNA pp format.</p>
<p>Optionally, one can specify structure and anchor constraints in these input files.</p>
<h2><a class="anchor" id="autotoc_md111"></a>
Output</h2>
<p>The program enumerates exactly matching local substructures (exact pattern matches = EPMs) and optionally chains them. It returns lists of chained and unchained matches, visualizations of the results and anchor constraints for alignment.</p>
<h1><a class="anchor" id="autotoc_md112"></a>
OPTIONS</h1>
<p><b>-h</b>, <b>&ndash;help</b></p>
<p>: Help</p>
<p><b>-V</b>, <b>&ndash;version</b></p>
<p>: Version info</p>
<p><b>-v</b>, <b>&ndash;verbose</b></p>
<p>: Verbose</p>
<p><b>-q</b>, <b>&ndash;quiet</b></p>
<p>: Quiet</p>
<h2><a class="anchor" id="autotoc_md113"></a>
Scoring parameters:</h2>
<p><b>&ndash;no-stacking</b></p>
<p>: Do not use stacking terms (otherwise needs stacking probs by RNAfold <b>-p2</b>)</p>
<p><b>&ndash;alpha_1=</b>&lt;factor&gt;(1)</p>
<p>: Multiplier for sequential score</p>
<p><b>&ndash;alpha_2=</b>&lt;factor&gt;(5)</p>
<p>: Multiplier for structural score</p>
<p><b>&ndash;alpha_3=</b>&lt;factor&gt;(5)</p>
<p>: Multiplier for stacking score, 0 means no stacking contribution</p>
<p><b>&ndash;struct-mismatch-score=</b>&lt;score&gt;(<b>-10</b>)</p>
<p>: Score for a structural mismatch (nucleotide mismatch in an arcmatch)</p>
<h2><a class="anchor" id="autotoc_md114"></a>
Heuristics for speed accuracy trade off:</h2>
<p><b>-p</b>, <b>&ndash;min-prob=</b>&lt;prob&gt;(0.01)</p>
<p>: Minimal probability</p>
<p><b>&ndash;max-bps-length-ratio=</b>&lt;factor&gt;(0.0)</p>
<p>: Maximal ratio of #base pairs divided by sequence length (default: no effect)</p>
<p><b>&ndash;max-uil-length-ratio=</b>&lt;factor&gt;(0.0)</p>
<p>: Maximal ratio of #unpaired bases in loops divided by sequence length (default: no effect)</p>
<p><b>&ndash;max-bpil-length-ratio=</b>&lt;factor&gt;(0.0)</p>
<p>: Maximal ratio of #base pairs in loops divided by loop length (default: no effect)</p>
<p><b>-D</b>, <b>&ndash;max-diff-am=</b>&lt;diff&gt;(30)</p>
<p>: Maximal difference for sizes of matched arcs</p>
<p><b>-d</b>, <b>&ndash;max-diff=</b>&lt;diff&gt;(<b>-1</b>)</p>
<p>: Maximal difference for alignment traces</p>
<p><b>&ndash;max-diff-at-am=</b>&lt;diff&gt;(<b>-1</b>)</p>
<p>: Maximal difference for alignment traces, only at arc match positions</p>
<p><b>&ndash;prob_unpaired_in_loop_threshold=</b>&lt;prob&gt;(0.01)</p>
<p>: Threshold for prob_unpaired_in_\ loop</p>
<p><b>&ndash;prob_basepair_in_loop_threshold=</b>&lt;prob&gt;(0.01)</p>
<p>: Threshold for prob_basepair_in_\ loop</p>
<h2><a class="anchor" id="autotoc_md115"></a>
Output:</h2>
<p><b>-p</b>, <b>&ndash;out-min-prob=</b>&lt;prob&gt;(0.0005)</p>
<p>: Minimal probability for output (min-prob overrides if smaller)</p>
<p><b>&ndash;output-ps</b></p>
<p>: Output best EPM chain as colored postscript</p>
<p><b>-a</b>, <b>&ndash;PS_fileA=</b>&lt;file&gt;()</p>
<p>: Postscript output file for sequence A</p>
<p><b>-b</b>, <b>&ndash;PS_fileB=</b>&lt;file&gt;()</p>
<p>: Postscript output file for sequence B</p>
<p><b>-o</b>, <b>&ndash;output-locarna=</b>&lt;file&gt;()</p>
<p>: Fasta file with anchor constraints for locarna</p>
<p><b>&ndash;output-anchor-pp=</b>&lt;fileroot&gt;()</p>
<p>: PP files &lt;fileroot&gt;_A.pp and &lt;fileroot&gt;_B.pp, merging input PPs and anchor constraints from chaining</p>
<p><b>&ndash;output-clustal=</b>&lt;file&gt;()</p>
<p>: Write file with chain as alignment in clustalw format</p>
<p><b>&ndash;output-epm-list=</b>&lt;file&gt;()</p>
<p>: A list of all found epms</p>
<p><b>&ndash;output-chained-epm-list=</b>&lt;file&gt;()</p>
<p>: A list of all EPMs that are present in the chain</p>
<p><b>&ndash;inexact-struct-match</b></p>
<p>: Allow inexact structure matches</p>
<p><b>&ndash;add-filter</b></p>
<p>: Apply an additional filter to enumerate only EPMs that are maximally extended (only inexact)</p>
<p><b>&ndash;no-chaining</b></p>
<p>: Do not use the chaining algorithm to find best overall chain</p>
<h2><a class="anchor" id="autotoc_md116"></a>
Suboptimal traceback:</h2>
<p><b>&ndash;subopt</b></p>
<p>: Use the suboptimal traceback</p>
<p><b>&ndash;diff-to-opt-score=</b>&lt;threshold&gt;(<b>-1</b>)</p>
<p>: Threshold for suboptimal traceback</p>
<p><b>&ndash;min-score=</b>&lt;score&gt;(90)</p>
<p>: Minimal score of a traced EPM</p>
<p><b>&ndash;number-of-EPMs=</b>&lt;threshold&gt;(100)</p>
<p>: Maximal number of EPMs for the suboptimal traceback</p>
<h2><a class="anchor" id="autotoc_md117"></a>
Constraints:</h2>
<p><b>&ndash;noLP</b></p>
<p>: use <b>&ndash;noLP</b> option for folding</p>
<p><b>&ndash;maxBPspan=</b>&lt;span&gt;(<b>-1</b>)</p>
<p>: Limit maximum base pair span (default=off)</p>
<p><b>&ndash;relaxed-anchors</b></p>
<p>: Relax anchor constraints (default=off)</p>
<h2><a class="anchor" id="autotoc_md118"></a>
Miscellaneous:</h2>
<p><b>&ndash;stopwatch</b></p>
<p>: Print run time information.</p>
<h2><a class="anchor" id="autotoc_md119"></a>
Input files:</h2>
<blockquote class="doxtable">
<p>The two input files &lt;Input 1&gt; and &lt;Input 2&gt; specify the input sequences in various, automatically detected formats. Accepted formats are: Fasta, Clustal, Stockholm LocARNA PP, ViennaRNA postscript dotplot. Unless in-loop</p>
<p>probabilities are provided (only possible in LocARNA PP), </p>
</blockquote>
<p>base pair</p>
<p>: probabilities are computed by partition function folding. </p><pre class="fragment">Clustal, Stockholm, and PP input can contain constraints.
</pre> <h1><a class="anchor" id="autotoc_md120"></a>
AVAILABILITY</h1>
<p>The latest LocARNA package release is available online at at Github <a href="https://github.com/s-will/LocARNA">https://github.com/s-will/LocARNA</a> and <a href="http://www.bioinf.uni-freiburg.de/Software/LocARNA/">http://www.bioinf.uni-freiburg.de/Software/LocARNA/</a></p>
<h1><a class="anchor" id="autotoc_md121"></a>
REFERENCES</h1>
<p>Christina Otto, Mathias Mohl, Steffen Heyne, Mika Amit, Gad M. Landau, Rolf Backofen, and Sebastian Will. ExpaRNA-P: simultaneous exact pattern matching and folding of RNAs. BMC Bioinformatics, 15:404, 2014. doi:10.1186/s12859-014-0404-0</p>
<h1><a class="anchor" id="autotoc_md122"></a>
EXAMPLES</h1>
<p>Essentially exparna_p is called in the same way as locarna and the other pairwise alignemnt tools of the package. Please refer to their documentation. Note that exparna is usually called with single sequences rather than alignments.</p>
<h1><a class="anchor" id="autotoc_md123"></a>
AUTHOR</h1>
<p>This man page is written and maintained by Sebastian Will it is part of the LocARNA package.</p>
<p>The exparna_p tool and the library classes for strong ensemble-based sparsification were written by Christina Otto.</p>
<h1><a class="anchor" id="autotoc_md124"></a>
REPORTING BUGS</h1>
<p>Report bugs to &lt;schmiedc (at) informatik.uni-freiburg.de&gt;.</p>
<h1><a class="anchor" id="autotoc_md125"></a>
COPYRIGHT</h1>
<p>Copyright 2005- Christina Otto, Sebastian Will. The LocARNA package is released under GNU Public License v3.0</p>
<h1><a class="anchor" id="autotoc_md126"></a>
SEE ALSO</h1>
<p>The LocARNA PP 2.0 format is described online at <a href="http://www.bioinf.uni-freiburg.de/Software/LocARNA/PP/">http://www.bioinf.uni-freiburg.de/Software/LocARNA/PP/</a> </p>
</div></div><!-- contents -->
</div><!-- PageDoc -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Wed Mar 30 2022 08:25:12 for LocARNA-2.0.0RC10 by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.1
</small></address>
</body>
</html>
