<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>LocARNA-2.0.0RC10: sparse</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">LocARNA-2.0.0RC10
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="PageDoc"><div class="header">
  <div class="headertitle">
<div class="title">sparse </div>  </div>
</div><!--header-->
<div class="contents">
<div class="toc"><h3>Table of Contents</h3>
<ul><li class="level1"><a href="#autotoc_md65">sparse</a></li>
<li class="level1"><a href="#autotoc_md66">NAME</a></li>
<li class="level1"><a href="#autotoc_md67">DESCRIPTION</a><ul><li class="level2"><a href="#autotoc_md68">Input, Constraints, and Output</a></li>
</ul>
</li>
<li class="level1"><a href="#autotoc_md69">OPTIONS</a><ul><li class="level2"><a href="#autotoc_md70">Scoring parameters:</a></li>
<li class="level2"><a href="#autotoc_md71">Partition function representation (for sequence envelopes):</a></li>
<li class="level2"><a href="#autotoc_md72">Controlling_output:</a></li>
<li class="level2"><a href="#autotoc_md73">Heuristics for speed accuracy trade off:</a></li>
<li class="level2"><a href="#autotoc_md74">MEA score:</a></li>
<li class="level2"><a href="#autotoc_md75">Constraints:</a></li>
<li class="level2"><a href="#autotoc_md76">Input files:</a></li>
</ul>
</li>
<li class="level1"><a href="#autotoc_md77">AVAILABILITY</a></li>
<li class="level1"><a href="#autotoc_md78">COPYING (LICENSE)</a></li>
<li class="level1"><a href="#autotoc_md79">REFERENCES</a></li>
<li class="level1"><a href="#autotoc_md80">AUTHOR</a></li>
<li class="level1"><a href="#autotoc_md81">REPORTING BUGS</a></li>
<li class="level1"><a href="#autotoc_md82">SEE ALSO</a></li>
</ul>
</div>
<div class="textblock"><h1><a class="anchor" id="autotoc_md65"></a>
sparse</h1>
<h1><a class="anchor" id="autotoc_md66"></a>
NAME</h1>
<p>sparse - manual page for sparse (<a class="el" href="namespaceLocARNA.html">LocARNA</a> 2.0.0RC10)</p>
<h1><a class="anchor" id="autotoc_md67"></a>
DESCRIPTION</h1>
<p>sparse - fast pairwise fast alignment of RNAs.</p>
<p>USAGE: sparse [options] &lt;Input 1&gt; &lt;Input 2&gt;</p>
<p>sparse is an even faster, stronger sparsified pairwise alignment algorithm than locarna. Like locarna, it performs fast simultaneous folding and alignment based on two RNA sequences (or alignments). In addition to the filtering of considered base pairs by their probabilities, it filters by conditional probabilities of bases and base pairs in their enclosing loops.</p>
<h2><a class="anchor" id="autotoc_md68"></a>
Input, Constraints, and Output</h2>
<p>The usage, input, constraints-specifications and output are analogous to locarna; please refer to the help or man page of locarna for explanations and examples.</p>
<h1><a class="anchor" id="autotoc_md69"></a>
OPTIONS</h1>
<p><b>-h</b>, <b>&ndash;help</b></p>
<p>: Print this help.</p>
<p><b>&ndash;galaxy-xml</b></p>
<p>: Print galaxy xml wrapper.</p>
<p><b>-V</b>, <b>&ndash;version</b></p>
<p>: Print only version string.</p>
<p><b>-v</b>, <b>&ndash;verbose</b></p>
<p>: Be verbose. Prints input parameters, sequences and size information.</p>
<p><b>-q</b>, <b>&ndash;quiet</b></p>
<p>: Be quiet.</p>
<h2><a class="anchor" id="autotoc_md70"></a>
Scoring parameters:</h2>
<p><b>-i</b>, <b>&ndash;indel=</b>&lt;score&gt;(<b>-150</b>)</p>
<p>: Indel score. Score contribution of each single base insertion or deletion. Indel opening score and indel score define the affine scoring of gaps.</p>
<p><b>-i</b>, <b>&ndash;indel-loop=</b>&lt;score&gt;(<b>-300</b>)</p>
<p>: Score for insertions and deletions of loops per base</p>
<p><b>&ndash;indel-opening=</b>&lt;score&gt;(<b>-750</b>)</p>
<p>: Indel opening score. Score contribution of opening an insertion or deletion, i.e. score for a consecutive run of deletions or insertions. Indel opening score and indel score define the affine scoring of gaps.</p>
<p><b>&ndash;indel-opening-loop=</b>&lt;score&gt;(<b>-900</b>)</p>
<p>: Opening score for insertions and deletions of loops</p>
<p><b>&ndash;ribosum-file=</b>&lt;f&gt;(RIBOSUM85_60)</p>
<p>: File specifying the Ribosum base and base-pair similarities. [default: use RIBOSUM85_60 without requiring a Ribosum file.]</p>
<p><b>&ndash;use-ribosum=</b>&lt;bool&gt;(true)</p>
<p>: Use ribosum scores for scoring base matches and base pair matches; note that tau=0 suppresses any effect on the latter.</p>
<p><b>-m</b>, <b>&ndash;match=</b>&lt;score&gt;(50)</p>
<p>: Set score contribution of a base match (unless ribosum scoring).</p>
<p><b>-M</b>, <b>&ndash;mismatch=</b>&lt;score&gt;(0)</p>
<p>: Set score contribution of a base mismatch (unless ribosum scoring).</p>
<p><b>&ndash;unpaired-penalty=</b>&lt;score&gt;(0)</p>
<p>: Penalty for unpaired bases</p>
<p><b>-s</b>, <b>&ndash;struct-weight=</b>&lt;score&gt;(200)</p>
<p>: Maximal weight of 1/2 arc match. Balances structure vs. sequence score contributions.</p>
<p><b>-e</b>, <b>&ndash;exp-prob=</b>&lt;prob&gt;</p>
<p>: Expected base pair probability. Used as background probability for base pair scoring [default: calculated from sequence length].</p>
<p><b>-t</b>, <b>&ndash;tau=</b>&lt;factor&gt;(100)</p>
<p>: Tau factor. Factor for contribution of sequence similarity in an arc match (in percent). tau=0 does not penalize any sequence information including compensatory mutations at arc matches, while tau=100 scores sequence similarity at ends of base matches (if a scoring matrix like ribosum is used, this adds the contributions for base pair match from the matrix). [default tau=0!]</p>
<p><b>-E</b>, <b>&ndash;exclusion=</b>&lt;score&gt;(0)</p>
<p>: Score contribution per exclusion in structure local alignment. Set to zero for unrestricted structure locality.</p>
<p><b>&ndash;stacking</b></p>
<p>: Use stacking terms (requires stack-probs by RNAfold <b>-p2</b>)</p>
<p><b>&ndash;new-stacking</b></p>
<p>: Use new stacking terms (requires stack-probs by RNAfold <b>-p2</b>)</p>
<h2><a class="anchor" id="autotoc_md71"></a>
Partition function representation (for sequence envelopes):</h2>
<p><b>&ndash;extended-pf</b></p>
<p>: Use extended precision for the computation of sequence envelopes. This enables handling significantly larger instances. [default]</p>
<p><b>&ndash;quad-pf</b></p>
<p>: Use quad precision for partition function values. Even more precision than extended pf, but usually much slower (overrides extended-pf).</p>
<h2><a class="anchor" id="autotoc_md72"></a>
Controlling_output:</h2>
<p><b>-w</b>, <b>&ndash;width=</b>&lt;columns&gt;(120)</p>
<p>: Width of alignment output.</p>
<p><b>&ndash;clustal=</b>&lt;file&gt;</p>
<p>: Write alignment in ClustalW (aln) format to given file.</p>
<p><b>&ndash;stockholm=</b>&lt;file&gt;</p>
<p>: Write alignment Stockholm format to given file.</p>
<p><b>&ndash;pp=</b>&lt;file&gt;</p>
<p>: Write alignment in PP format to given file.</p>
<p><b>&ndash;alifold-consensus-dp</b></p>
<p>: Compute consensus dot plot by alifold (warning: this may fail for long sequences).</p>
<p><b>&ndash;consensus-structure=</b>&lt;type&gt;(alifold)</p>
<p>: Type of consensus structures written to screen and stockholm output [alifold|mea|none] (default: none).</p>
<p><b>&ndash;consensus-gamma=</b>&lt;float&gt;(1.0)</p>
<p>: Base pair weight for mea consensus computation. For MEA, base pairs are scored by their pair probability times 2 gamma; unpaired bases, by their unpaired probability.</p>
<p><b>-L</b>, <b>&ndash;local-output</b></p>
<p>: Output only local sub-alignment (to std out).</p>
<p><b>&ndash;local-file-output</b></p>
<p>: Write only local sub-alignment to output files.</p>
<p><b>-P</b>, <b>&ndash;pos-output</b></p>
<p>: Output only local sub-alignment positions.</p>
<p><b>&ndash;write-structure</b></p>
<p>: Write guidance structure in output.</p>
<p><b>&ndash;special-gap-symbols</b></p>
<p>: Special distinct gap symbols for loop gaps or gaps caused by sparsification</p>
<p><b>&ndash;stopwatch</b></p>
<p>: Print run time informations.</p>
<h2><a class="anchor" id="autotoc_md73"></a>
Heuristics for speed accuracy trade off:</h2>
<p><b>-p</b>, <b>&ndash;min-prob=</b>&lt;prob&gt;(0.001)</p>
<p>: Minimal probability. Only base pairs of at least this probability are taken into account.</p>
<p><b>&ndash;prob-unpaired-in-loop-threshold=</b>&lt;threshold&gt;(0.00005)</p>
<p>: Threshold for prob_unpaired_in_loop</p>
<p><b>&ndash;prob-basepair-in-loop-threshold=</b>&lt;threshold&gt;(0.0001)</p>
<p>: Threshold for prob_basepair_in_loop</p>
<p><b>&ndash;max-bps-length-ratio=</b>&lt;factor&gt;(0.0)</p>
<p>: Maximal ratio of #base pairs divided by sequence length. This serves as a second filter on the "significant" base pairs. [default: 0.0 = no effect].</p>
<p><b>&ndash;max-uil-length-ratio=</b>&lt;factor&gt;(0.0)</p>
<p>: Maximal ratio of #unpaired bases in loops divided by sequence length (def: no effect)</p>
<p><b>&ndash;max-bpil-length-ratio=</b>&lt;factor&gt;(0.0)</p>
<p>: Maximal ratio of #base pairs in loops divided by loop length (def: no effect)</p>
<p><b>-D</b>, <b>&ndash;max-diff-am=</b>&lt;diff&gt;(<b>-1</b>)</p>
<p>: Maximal difference for sizes of matched arcs. [-1=off]</p>
<p><b>-d</b>, <b>&ndash;max-diff=</b>&lt;diff&gt;(<b>-1</b>)</p>
<p>: Maximal difference for positions of alignment traces (and aligned bases). [-1=off]</p>
<p><b>&ndash;max-diff-at-am=</b>&lt;diff&gt;(<b>-1</b>)</p>
<p>: Maximal difference for positions of alignment traces at arc match ends. [-1=off]</p>
<p><b>&ndash;max-diff-aln=</b>&lt;aln file&gt;()</p>
<p>: Maximal difference relative to given alignment (file in clustalw format)</p>
<p><b>&ndash;max-diff-pw-aln=</b>&lt;alignment&gt;()</p>
<p>: Maximal difference relative to given alignment (string, delim=AMPERSAND)</p>
<p><b>&ndash;max-diff-relax</b></p>
<p>: Relax deviation constraints in multiple aligmnent</p>
<p><b>&ndash;min-trace-probability=</b>&lt;probability&gt;(1e-5)</p>
<p>: Minimal sequence alignment probability of potential traces (probability-based sequence alignment envelope) [default=1e-4].</p>
<h2><a class="anchor" id="autotoc_md74"></a>
MEA score:</h2>
<p><b>&ndash;mea-alignment</b></p>
<p>: Perform maximum expected accuracy alignment (instead of using the default similarity scoring).</p>
<p><b>&ndash;match-prob-method=</b>&lt;int&gt;(0)</p>
<p>: Select method for computing sequence-based base match probablities (to be used for mea-type alignment scores). Methods: 1=probcons-style from HMM, 2=probalign-style from PFs, 3=from PFs, local</p>
<p><b>&ndash;probcons-file=</b>&lt;file&gt;</p>
<p>: Read parameters for probcons-like calculation of match probabilities from probcons parameter file.</p>
<p><b>&ndash;temperature-alipf=</b>&lt;int&gt;(300)</p>
<p>: Temperature for the */sequence* alignment/ partition functions used by the probcons-like sequence-based match/trace probability computation (this temperature is different from the \'physical\' temperature of RNA folding!).</p>
<p><b>&ndash;pf-struct-weight=</b>&lt;weight&gt;(200)</p>
<p>: Structure weight in PF computations (for the computation of sequence-based match probabilties from partition functions).</p>
<p><b>&ndash;mea-gapcost</b></p>
<p>: Use gap cost in mea alignment</p>
<p><b>&ndash;mea-alpha=</b>&lt;weight&gt;(0)</p>
<p>: Weight alpha for MEA</p>
<p><b>&ndash;mea-beta=</b>&lt;weight&gt;(200)</p>
<p>: Weight beta for MEA</p>
<p><b>&ndash;mea-gamma=</b>&lt;weight&gt;(100)</p>
<p>: Weight gamma for MEA</p>
<p><b>&ndash;probability-scale=</b>&lt;scale&gt;(10000)</p>
<p>: Scale for probabilities/resolution of mea score</p>
<p><b>&ndash;write-match-probs=</b>&lt;file&gt;</p>
<p>: Write match probs to file (don\'t align!).</p>
<p><b>&ndash;read-match-probs=</b>&lt;file&gt;</p>
<p>: Read match probabilities from file.</p>
<p><b>&ndash;write-arcmatch-scores=</b>&lt;file&gt;</p>
<p>: Write arcmatch scores (don\'t align!)</p>
<p><b>&ndash;read-arcmatch-scores=</b>&lt;file&gt;</p>
<p>: Read arcmatch scores.</p>
<p><b>&ndash;read-arcmatch-probs=</b>&lt;file&gt;</p>
<p>: Read arcmatch probabilities (weighted by factor mea_beta/100)</p>
<h2><a class="anchor" id="autotoc_md75"></a>
Constraints:</h2>
<p><b>&ndash;noLP</b></p>
<p>: Disallow lonely pairs in prediction and alignment.</p>
<p><b>&ndash;maxBPspan=</b>&lt;span&gt;(<b>-1</b>)</p>
<p>: Limit maximum base pair span [default=off].</p>
<p><b>&ndash;relaxed-anchors</b></p>
<p>: Use relaxed semantics of anchor constraints [default=strict semantics].</p>
<h2><a class="anchor" id="autotoc_md76"></a>
Input files:</h2>
<blockquote class="doxtable">
<p>The tool is called with two input files &lt;Input 1&gt; and &lt;Input 2&gt;, which specify the two input sequences or input alignments. Different input formats (Fasta, Clustal, Stockholm, <a class="el" href="namespaceLocARNA.html">LocARNA</a> PP, ViennaRNA postscript dotplots) are accepted and automatically recognized (by file content); the two input files can be in different formats. Extended variants of the Clustal and Stockholm formats enable specifying anchor and structure constraints. </p>
</blockquote>
<h1><a class="anchor" id="autotoc_md77"></a>
AVAILABILITY</h1>
<p>The latest <a class="el" href="namespaceLocARNA.html">LocARNA</a> package release is available online at at Github <a href="https://github.com/s-will/LocARNA">https://github.com/s-will/LocARNA</a> and <a href="http://www.bioinf.uni-freiburg.de/Software/LocARNA/">http://www.bioinf.uni-freiburg.de/Software/LocARNA/</a></p>
<h1><a class="anchor" id="autotoc_md78"></a>
COPYING (LICENSE)</h1>
<p>Copyright 2005- Milad Miladi, Sebastian Will, Christina Otto. The <a class="el" href="namespaceLocARNA.html">LocARNA</a> package is released under GNU Public License v3.0</p>
<h1><a class="anchor" id="autotoc_md79"></a>
REFERENCES</h1>
<p>Sebastian Will, Christina Otto, Milad Miladi, Mathias M??hl, and Rolf Backofen. SPARSE: quadratic time simultaneous alignment and folding of RNAs without sequence-based heuristics. Bioinformatics 31 (15): 2489-2496, 2015. doi:10.1093/bioinformatics/btv185</p>
<h1><a class="anchor" id="autotoc_md80"></a>
AUTHOR</h1>
<p>This man page is written and maintained by Sebastian Will it is part of the <a class="el" href="namespaceLocARNA.html">LocARNA</a> package.</p>
<p>The sparse tool and sparse alignment algorithm is written by Milad Miladi. Library classes for strong ensemble-based sparsification are written by Christina Otto.</p>
<h1><a class="anchor" id="autotoc_md81"></a>
REPORTING BUGS</h1>
<p>Report bugs to &lt;miladim (at) informatik.uni-freiburg.de&gt;.</p>
<h1><a class="anchor" id="autotoc_md82"></a>
SEE ALSO</h1>
<p>The <a class="el" href="namespaceLocARNA.html">LocARNA</a> PP 2.0 format is described online at <a href="http://www.bioinf.uni-freiburg.de/Software/LocARNA/PP/">http://www.bioinf.uni-freiburg.de/Software/LocARNA/PP/</a> </p>
</div></div><!-- contents -->
</div><!-- PageDoc -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Tue Mar 29 2022 16:02:21 for LocARNA-2.0.0RC10 by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.1
</small></address>
</body>
</html>
