<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>LocARNA-2.0.0RC10: mlocarna</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">LocARNA-2.0.0RC10
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="PageDoc"><div class="header">
  <div class="headertitle">
<div class="title">mlocarna </div>  </div>
</div><!--header-->
<div class="contents">
<div class="toc"><h3>Table of Contents</h3>
<ul><li class="level1"><a href="#autotoc_md13">mlocarna</a></li>
<li class="level1"><a href="#autotoc_md14">NAME</a></li>
<li class="level1"><a href="#autotoc_md15">SYNOPSIS</a></li>
<li class="level1"><a href="#autotoc_md16">DESCRIPTION</a></li>
<li class="level1"><a href="#autotoc_md17">OPTIONS</a><ul><li class="level2"><a href="#autotoc_md18">Load configurations from file</a></li>
<li class="level2"><a href="#autotoc_md19">Major alignment modes</a></li>
<li class="level2"><a href="#autotoc_md20">Controlling Output</a></li>
<li class="level2"><a href="#autotoc_md21">Locality</a></li>
<li class="level2"><a href="#autotoc_md22">Pairwise alignment and scoring</a></li>
<li class="level2"><a href="#autotoc_md23">Alignment heuristics</a></li>
<li class="level2"><a href="#autotoc_md24">Low-level selection of pairwise alignment tools and options</a></li>
<li class="level2"><a href="#autotoc_md25">Controlling the guide tree construction</a></li>
<li class="level2"><a href="#autotoc_md26">Controlling multiple alignment construction</a></li>
<li class="level2"><a href="#autotoc_md27">Further options for probabilistic mode</a></li>
<li class="level2"><a href="#autotoc_md28">Miscallaneous modes of operation</a></li>
<li class="level2"><a href="#autotoc_md29">Constraints</a></li>
<li class="level2"><a href="#autotoc_md30">Rna folding (RNAfold/RNAplfold)</a></li>
<li class="level2"><a href="#autotoc_md31">Multithreading</a></li>
<li class="level2"><a href="#autotoc_md32">Getting Help</a></li>
</ul>
</li>
<li class="level1"><a href="#autotoc_md33">EXAMPLES</a><ul><li class="level2"><a href="#autotoc_md34">Calling mlocarna</a></li>
<li class="level2"><a href="#autotoc_md35">Use of constraints</a></li>
<li class="level2"><a href="#autotoc_md36">Realignment</a></li>
</ul>
</li>
<li class="level1"><a href="#autotoc_md37">AUTHORS</a></li>
<li class="level1"><a href="#autotoc_md38">ONLINE INFORMATION</a></li>
<li class="level1"><a href="#autotoc_md39">REFERENCES</a></li>
</ul>
</div>
<div class="textblock"><h1><a class="anchor" id="autotoc_md13"></a>
mlocarna</h1>
<h1><a class="anchor" id="autotoc_md14"></a>
NAME</h1>
<p><b>MLocARNA</b> - multiple alignment of RNA</p>
<h1><a class="anchor" id="autotoc_md15"></a>
SYNOPSIS</h1>
<p>mlocarna [options] &lt;fasta file&gt;</p>
<h1><a class="anchor" id="autotoc_md16"></a>
DESCRIPTION</h1>
<p><b>MLocarna</b> computes a multiple sequence-structure alignment of RNA sequences. The structure of these sequences does not have to be known but is inferred based in simultaneous alignment and folding.</p>
<p>Generally, mlocarna takes multiple sequences as input, given in a fasta file. The fasta file can be extended to specify structure and anchor constraints that respectively control the possible foldings and possible alignments. The main outcome is a multiple alignment together with a consensus structure.</p>
<p>Technically, mlocarna works as front end to the pairwise alignment tools locarna, locarna_p, and sparse (and even carna), which are employed to construct the multiple alignment progressively.</p>
<p>Going beyond the basic progressive alignment scheme, Mlocarna implements probabilistic consistency transformation and iterative alignment, which are available in probabilistic mode. Moreover, the <a class="el" href="namespaceLocARNA.html">LocARNA</a> package provides an alternative multiple alignment tool "locarnate", which generates alignments based on T-Coffee using (non-probabilistic) consistency transformation.</p>
<h1><a class="anchor" id="autotoc_md17"></a>
OPTIONS</h1>
<h2><a class="anchor" id="autotoc_md18"></a>
Load configurations from file</h2>
<ul>
<li><p class="startli"><b>&ndash;configure</b>=file</p>
<p class="startli">Load a parameter set from a configuration file of options and option value pairs. This enables specifying (sets of) default parameters for mlocarna, which can still be modified by other options to mlocarna. Command line arguments always take precedence over this configuration. Options are specified as single entries per line; option value pairs, like option: value. Whitespace and '#'-prefixed comments are ignored.</p>
</li>
</ul>
<h2><a class="anchor" id="autotoc_md19"></a>
Major alignment modes</h2>
<p>By default, mlocarna performs progressive alignment, where the progressive alignment steps are computed by the pairwise aligner locarna based on sequences and dot plots (RNAfold -p); subsequently, partial alignmetns and their consensus dot plots.</p>
<ul>
<li><p class="startli"><b>&ndash;probabilistic</b></p>
<p class="startli">In probabilistic mode, mlocarna scores alignments using match probabilities that are computed by a partition function approach [tech. details: the probability computation is implemented in locarna_p; the probability-based scoring is performed by locarna in mea mode]. This enables mlocarna to consistency-transform the probabilities (option &ndash;consistency-transform) and to compute reliabilities. The tool reliability-profile.pl is provided to visualize reliability profiles. Reliabilities can also be used for iterating the alignment with reliably aligned base pairs as structural constraints (option &ndash;it-reliable-structure).</p>
</li>
<li><p class="startli"><b>&ndash;sparse</b></p>
<p class="startli">Apply the sparsified alignment algorithm SPARSE for all pairwise alignments (instead of the default pairwise aligner locarna). SPARSE supports stronger sparsification for faster alignment computation and increases the structure prediction capabilities over locarna.</p>
</li>
</ul>
<h2><a class="anchor" id="autotoc_md20"></a>
Controlling Output</h2>
<ul>
<li><p class="startli"><b>&ndash;tgtdir</b></p>
<p class="startli">Target directory. All output files are written to this directory. Per default the target directory is generated from the input filename by replacing suffix fa by (or appending) out.</p>
</li>
<li><p class="startli"><b>-v, &ndash;verbose</b></p>
<p class="startli">Turn on verbose ouput. Shows progress of computation of all-2-all pairwise alignments for guide tree computation; shows intermediary alignments during the progressive alignment computation.</p>
</li>
<li><p class="startli"><b>&ndash;moreverbose</b></p>
<p class="startli">Be even more verbose: additionally shows parameters for the pairwise aligner; moreover, the calls and output of the RNA base pair probability computations as well as the pairwise aligner during progressive alignment.</p>
</li>
<li><p class="startli"><b>-q, &ndash;quiet</b></p>
<p class="startli">Be quiet.</p>
</li>
<li><p class="startli"><b>&ndash;keep-sequence-order</b></p>
<p class="startli">Preserve sequence order of the input in the final alignment. Affects output to stdout and results/result.aln.</p>
</li>
<li><p class="startli"><b>&ndash;stockholm</b></p>
<p class="startli">Write STOCKHOLM files of all final and intermediate alignments (in addition to CLUSTALW files).</p>
</li>
<li><p class="startli"><b>&ndash;consensus-structure</b></p>
<p class="startli">Type of consensus structures written to screen and stockholm output [alifold|mea|none] (default: none). Unless explicitly specified, the option alifold-consensus-dp implicitly sets this to alifold.</p>
</li>
<li><p class="startli"><b>-w, &ndash;width</b>=columns (120)</p>
<p class="startli">Output width for sequences in clustal-like and stockholm output; note that the clustalw standard format requires 60 or less.</p>
</li>
<li><p class="startli"><b>&ndash;write-structure</b></p>
<p class="startli">Write guidance structure in output to stdout. This provides some insight into the influence of structure into the generated pairwise alignments. The guidance structure shows the base pairs 'predicted' by each pairwise locarna (or sparse) alignment. These structures should not be mistaken as predicted consensus structures of multiple alignments. Consensus structures can be more adequately derived from the multiple alignment. For this reason, mlocarna reports the consensus structure by RNAalifold.</p>
</li>
</ul>
<h2><a class="anchor" id="autotoc_md21"></a>
Locality</h2>
<ul>
<li><p class="startli"><b>&ndash;free-endgaps</b></p>
<p class="startli">Allow free endgaps. (Corresponds to pairwise locarna option &ndash;free-endgaps "++++".)</p>
</li>
<li><p class="startli"><b>&ndash;free-endgaps-3</b></p>
<p class="startli">Allow free endgaps 3'.</p>
</li>
<li><p class="startli"><b>&ndash;free-endgaps-5</b></p>
<p class="startli">Allow free endgaps 5'.</p>
</li>
<li><p class="startli"><b>&ndash;sequ-local</b>=bool (false)</p>
<p class="startli">Turns on/off sequence locality [def=off]. Sequence locality refers to the usual form of local alignment. If on, mlocarna bases all calculations on local pairwise alignments, which determin the best alignments of subsequences (disregarding dissimilar starts and ends). Note that truely local structure alignments as well as local multiple alignments are still a matter of research; so don't expect perfect results in all instances.</p>
</li>
<li><p class="startli"><b>&ndash;struct-local</b>=bool (false)</p>
<p class="startli">Turns on/off structure locality [def=off]. Structural locality enables skipping entire substructures in alignments. In pairwise alignments, this allows one exclusion of some subsequence in each loop; thus, guaranteeing that the (structure locally) aligned parts of the sequences are always connected w.r.t. the predicted structure but not necessarily consecutive in the sequence. Structure locality does not imply sequence locality, but rather the two concepts are orthogonal.</p>
</li>
<li><p class="startli"><b>&ndash;penalized</b>=score</p>
<p class="startli">Variant of sequence local alignment (cf. &ndash;sequ-local), where the specified penalty score is subtracted for each base in the local alignment. [Experimental]</p>
</li>
</ul>
<h2><a class="anchor" id="autotoc_md22"></a>
Pairwise alignment and scoring</h2>
<ul>
<li><p class="startli"><b>&ndash;indel=score</b> (-150)</p>
<p class="startli">Score of each single base insertion or deletion.</p>
</li>
<li><p class="startli"><b>&ndash;indel-opening</b>=score (-750)</p>
<p class="startli">Score of opening an insertion or deletion, i.e. score for a consecutive run of deletions or insertions. Indel opening score and indel score define the affine scoring of gaps.</p>
</li>
<li><p class="startli"><b>-m, &ndash;match</b>=score (50)</p>
<p class="startli">Score of a base match (unless ribosum-based scoring)</p>
</li>
<li><p class="startli"><b>-M, &ndash;mismatch</b>=score (0)</p>
<p class="startli">Score of a base mismatch (unless ribosum-based scoring)</p>
</li>
<li><p class="startli"><b>&ndash;use-ribosum</b>=bool (true)</p>
<p class="startli">Use ribosum scores for scoring base matches and base pair matches; note that tau=0 suppresses any effect on the latter.</p>
</li>
<li><p class="startli"><b>&ndash;ribosum-file</b>=file</p>
<p class="startli">File specifying the Ribosum base and base-pair similarities. [default: use RIBOSUM85_60 without requiring a Ribosum file.]</p>
</li>
<li><p class="startli"><b>-s, &ndash;struct-weight</b>=score (200)</p>
<p class="startli">Maximum weight of one predicted arc, aka base pair. Note that this means that the maximum weight of an arc match is twice as high. The maximum weight is assigned to base pairs with (almost) probability 1 in the dot plot; less probable base pairs receive gradually degrading scores. The struct-weight factor balances the score contribution from structure to the score contribution from base similarity scores (e.g. ribosum scores).</p>
</li>
<li><p class="startli"><b>-e, &ndash;exp-prob</b>=prob</p>
<p class="startli">Expected probability of a base pair.</p>
</li>
<li><p class="startli"><b>-t, &ndash;tau</b>=factor (0)</p>
<p class="startli">Tau factor in percent. The tau factor controls the contribution of sequence-dependent scores to the score of arc matches.</p>
</li>
<li><p class="startli"><b>-E, &ndash;exclusion</b>=&lt;score&gt; (0)</p>
<p class="startli">Weight of an exclusion, i.e. an ommitted subsequence in a loop, which applies only to structural local alignment.</p>
</li>
<li><p class="startli"><b>&ndash;stacking</b></p>
<p class="startli">Use stacking terms. In this case, stacked arcs are scored based on conditional probabilities (conditioned by their stacked inner arc) rather than unconditioned base pair probabilities. [Experimental]</p>
</li>
<li><p class="startli"><b>&ndash;new-stacking</b></p>
<p class="startli">Use new stacking terms; cf. &ndash;stacking. These terms directly award bonuses to stacking. [Experimental]</p>
</li>
</ul>
<h2><a class="anchor" id="autotoc_md23"></a>
Alignment heuristics</h2>
<p>Several parameters are available to speed up the pairwise alignment computations heuristically. Choosing these parameters reasonably is necessary to achieve good trade-off between speed and accuracy, especially for large alignment instances.</p>
<ul>
<li><p class="startli"><b>-p, &ndash;min-prob</b>=probability (0.001)</p>
<p class="startli">Minimum base pair / arc probability. Arc with lower probability in the input RNA structure ensembles are ignored.</p>
</li>
<li><p class="startli"><b>-P, &ndash;tree-min-prob</b>=probability</p>
<p class="startli">Minimal prob for constructing guide tree. This probability can be set separately for the all-2-all comparison for constructing the guide tree and the progressive/iterative alignment steps.</p>
</li>
<li><p class="startli"><b>&ndash;max-bps-length-ratio</b>=factor (0.0)</p>
<p class="startli">Maximal ratio of the number of base pairs divided by sequence length (default: no effect)</p>
</li>
<li><p class="startli"><b>-D, &ndash;max-diff-am</b>=difference</p>
<p class="startli">Maximal difference for lengths of matched arcs. Two arcs that have a higher difference of their lengths are ignored. This speeds up the alignment, since less arc comparisons (i.e. less DP matrices) have to be computed. [def: off/-1]</p>
</li>
<li><p class="startli"><b>-d, &ndash;max-diff</b>=difference</p>
<p class="startli">Maximal difference of the positions of any two bases that are considered to be aligned. Bases with higher difference are generally not aligned. This allows banding of the DP matrices and thus can result in high speed ups. Note that the semantic changes in the context of a reference alignment specified with max-diff-aln. Then, the difference to the reference alignment is restricted. [def: off/-1]</p>
</li>
<li><p class="startli"><b>&ndash;max-diff-at-am</b>=difference</p>
<p class="startli">Same restriction as max-diff but only at the ends of arcs in arc matches. [def: off/-1]</p>
</li>
<li><p class="startli"><b>&ndash;min-trace-probability</b>=probability</p>
<p class="startli">Minimal sequence alignment probability of potential traces (probability-based sequence alignment envelope) [default=1e-4, moderate filter].</p>
</li>
<li><p class="startli"><b>&ndash;max-diff-aln</b>=file</p>
<p class="startli">Computes "realignment" in the environment of the given reference alignment (file in clustalw format) by constraining the maximum difference to this reference (controlled by &ndash;max-diff). The input sequences (and their names) have to be identical to these alignment sequences; however the alignment is allowed to contain extra sequences, which are ignored. In combination with option &ndash;realign, the reference alignment is taken from the (main) input file. In this case, the 'file' argument should be '.', but is ignored (with warning) otherwise.</p>
</li>
<li><p class="startli"><b>&ndash;max-diff-relax</b></p>
<p class="startli">Relax deviation constraints (cf. &ndash;max-diff-aln) in multiple aligmnent. This option is useful if the default strategy for realignment fails.</p>
</li>
<li><p class="startli"><b>-a, &ndash;min-am-prob</b>=probability (0.001)</p>
<p class="startli">Minimum arc-match probability (filters output of locarna-p)</p>
</li>
<li><p class="startli"><b>-b, &ndash;min-bm-prob</b>=probability (0.001)</p>
<p class="startli">Minimum base-match probability (filters output of locarna-p)</p>
</li>
</ul>
<h2><a class="anchor" id="autotoc_md24"></a>
Low-level selection of pairwise alignment tools and options</h2>
<ul>
<li><p class="startli"><b>&ndash;pw-aligner</b></p>
<p class="startli">Utilize the given tool for computing pairwise alignments (def=locarna).</p>
</li>
<li><p class="startli"><b>&ndash;pw-aligner-p</b>=tool</p>
<p class="startli">Utilize the given tool for computing partition function pairwise alignments (def=locarna_p).</p>
</li>
<li><p class="startli"><b>&ndash;pw-aligner-options</b></p>
<p class="startli">Additional option string for the pairwise alignment tool (def="").</p>
</li>
<li><p class="startli"><b>&ndash;pw-aligner-p-options</b></p>
<p class="startli">Additional option string for the partition function pairwise alignment tool (def="").</p>
</li>
</ul>
<h2><a class="anchor" id="autotoc_md25"></a>
Controlling the guide tree construction</h2>
<ul>
<li><p class="startli"><b>&ndash;treefile</b>=file</p>
<p class="startli">File with guide tree in NEWICK format. The given tree is used as guide tree for the progressive alignment. This saves the calculation of pairwise all-vs-all similarities and construction of the guide tree.</p>
</li>
<li><p class="startli"><b>&ndash;similarity-matrix</b>=file</p>
<p class="startli">File with similarity matrix. The similarities in the matrix are used to construct the guide tree for the progressive alignment. This saves the calculation of pairwise all-vs-all similarities.</p>
</li>
<li><p class="startli"><b>&ndash;score-lists</b></p>
<p class="startli">Construct the guide tree from pairwise scores in files scores* in the subdirectory scores of the target directory. The scores are typically precomputed, possibly in a distributed way, using &ndash;compute-pairwise-scores.</p>
</li>
<li><p class="startli"><b>&ndash;compute-pairwise-scores</b>=k/N</p>
<p class="startli">Compute only the pairwise alignments for the guide tree construction. Write scores to the file $tgtdir/scores/scores-$k and terminate. By computing only the k-th fraction of N parts, the option supports distributing the computation of the alignments. Before computing the pairwise scores, the dot plot files should be precomputed using &ndash;only-dps. (see also: &ndash;score-lists)</p>
</li>
<li><p class="startli"><b>&ndash;graphkernel</b></p>
<p class="startli">Use the graphkernel for constructing the guide tree.</p>
</li>
<li><p class="startli"><b>&ndash;svmsgdnspdk</b>=program</p>
<p class="startli">Specify the svmsgdnspdk program (potentially including path). Default: use "svmsgdnspdk" in path.</p>
</li>
<li><p class="startli"><b>&ndash;fasta2shrep</b>=program</p>
<p class="startli">Program "fasta2shrep" for generating graphs from the input sequences for use with the graph kernel guide tree generation (potentially including path). Default: use "fasta2shrep\_gspan.pl" in path.</p>
</li>
<li><p class="startli"><b>&ndash;fasta2shrep-options</b>=argument-string</p>
<p class="startli">Command line arguments for fasta2shrep. Default: "-wins 200 -shift 50
    \-stack -t 3 -M 3".</p>
</li>
</ul>
<h2><a class="anchor" id="autotoc_md26"></a>
Controlling multiple alignment construction</h2>
<ul>
<li><p class="startli"><b>&ndash;alifold-consensus-dp</b></p>
<p class="startli">Employs <b>RNAalifold -p</b> for generating consensus dotplot after each progressive alignment step. This replaces the default consensus dotplot computation, which averages over the input dot plots. This method should be used with care in combination with structural constraints, since it ignores them for all but the pairwise alignments of single sequences. Furthermore, note that it does not support <b>&ndash;stacking</b> or <b>&ndash;new-stacking</b>.</p>
</li>
<li><p class="startli"><b>&ndash;max-alignment-size</b>=size</p>
<p class="startli">Limit the maximum number of sequences that are aligned together by progressive alignment. This can be used to save unnecessary computations, when producing a clustering of the input RNAs rather than constructing a single multiple alignment. [default: no limit].</p>
</li>
<li><p class="startli"><b>&ndash;local-progressive</b></p>
<p class="startli">Align only the subalignment of locally aligned subsequences in subsequent steps of the progressive multiple alignment. Note: this is only effective if local alignment is turned on. (Default for sequence local alignment; turn off by <b>&ndash;global-progressive</b>)</p>
</li>
<li><p class="startli"><b>&ndash;global-progressive</b></p>
<p class="startli">Use alignments including "locality gaps" in subsequent steps of the progressive multiple alignment. Note: this is only effective if local alignment is turned on. (Opposite of <b>&ndash;local-progressive</b>)</p>
</li>
<li><p class="startli"><b>&ndash;consistency-transformation</b></p>
<p class="startli">Apply probabilistic consistency transformation (only possible in probabilistic mode).</p>
</li>
<li><p class="startli"><b>&ndash;iterate</b></p>
<p class="startli">Refine iteratively after progressive alignment. Currently, iterative refinement optimizes the SCI or RELIABILITY (not the locarna score)! Iterative refinement realigns all binary splits along the guide tree.</p>
</li>
<li><p class="startli"><b>&ndash;iterations</b>=number</p>
<p class="startli">Refine iteratively for given number of iterations (or stop at convergence).</p>
</li>
<li><p class="startli"><b>&ndash;it-reliable-structure</b>=number</p>
<p class="startli">Iterate alignment &lt;num&gt; times with reliable structure. This works only in probabilistic mode, when reliabilities can be computed.</p>
</li>
</ul>
<h2><a class="anchor" id="autotoc_md27"></a>
Further options for probabilistic mode</h2>
<ul>
<li><p class="startli"><b>&ndash;pf-only-basematch-probs</b></p>
<p class="startli">Use only base match probabilities (no base pair match probabilities).</p>
</li>
<li><b>&ndash;extended-pf</b> <pre class="fragment">  Use extended precision for partition function values. This increases
  run-time and space (less than 2x), however enables handling
  significantly larger instances.
</pre></li>
<li><b>&ndash;quad-pf</b> <pre class="fragment">  Use quad precision for partition function values. Even more precision
  than extended pf, but usually much slower (overrides extended-pf).
</pre></li>
<li><p class="startli"><b>&ndash;pf-scale=&lt;scale</b>&gt;</p>
<p class="startli">Scale of partition function; use for avoiding overflow in larger instances.</p>
</li>
<li><p class="startli"><b>&ndash;fast-mea</b></p>
<p class="startli">Compute base match probabilities using Gotoh PF-algorithm.</p>
</li>
<li><p class="startli"><b>&ndash;mea-alpha</b></p>
<p class="startli">Weight of unpaired probabilities in fast mea mode.</p>
</li>
<li><p class="startli"><b>&ndash;mea-beta</b></p>
<p class="startli">Weight of base pair match contribution in probabilistic mode.</p>
</li>
<li><p class="startli"><b>&ndash;mea-gamma</b></p>
<p class="startli">Reserved parameter for fast-mea mode.</p>
</li>
<li><p class="startli"><b>&ndash;mea-gapcost</b></p>
<p class="startli">Turn on gap penalties in probabilistic/mea mode (default: off).</p>
</li>
<li><p class="startli"><b>&ndash;write-probs</b> / <b>&ndash;no-write-probs</b></p>
<p class="startli">Write / don't write probabilities (of base matches and arc matches) to the target directory. Override by single options &ndash;(no-)write-bm-probs and &ndash;(no-)write-am-probs is possible. Use this to make the probability files available for post-processing. (default: don't write).</p>
</li>
<li><p class="startli"><b>&ndash;write-bm-probs</b> / <b>&ndash;no-write-bm-probs</b></p>
<p class="startli">Don't write / Write base match probabilities to files in target dir (default: don't write).</p>
</li>
<li><p class="startli"><b>&ndash;write-am-probs</b> / <b>&ndash;no-write-am-probs</b></p>
<p class="startli">Don't write / Write arc match probabilities to files in target dir (default: don't write).</p>
</li>
</ul>
<h2><a class="anchor" id="autotoc_md28"></a>
Miscallaneous modes of operation</h2>
<ul>
<li><p class="startli"><b>&ndash;realign</b></p>
<p class="startli">Realignment mode. In this mode, the input must be in clustal format and is interpreted as alignment of the input sequences; the sequences are obtained by removing all gap symbols. Moreover, the given alignment is set as reference alignment for &ndash;max-diff-aln. Structure and anchor constraints can be specified as consensus constraints in the input; constraints are specified as 'alignment strings' with names '#A1', '#S', or '#FS' for anchor, structure, or fixed structure constraints, respectively. Characters in the '#A1' anchor specification other than '-' and '.' constrain the aligned residues in the respective column to remain aligned (blanks are disallowed; annotations '#A2', '#A3', ... are ignored). The consensus structure constraint is equivalent to constraining each single sequence by the projection of the consensus constraint to the sequence (removing all base pairs with at least one gapped end).</p>
</li>
<li><p class="startli"><b>&ndash;dp-cache</b>=directory</p>
<p class="startli">Use directory &lt;dir&gt; as cache for dot plot or pp files (useful for avoiding multiple computation).</p>
</li>
<li><p class="startli"><b>&ndash;only-dps</b></p>
<p class="startli">Compute only the pair probability files / dot plots, don't align (useful for filling the dp-cache).</p>
</li>
<li><p class="startli"><b>&ndash;evaluate</b>=file</p>
<p class="startli">Evaluate the given multiple alignment (clustalw aln format, or use --eval-fasta). This requires that probailities are already computed (mlocarna &ndash;probabilistic) and present in the target directory (&ndash;tgtdir).</p>
</li>
<li><p class="startli"><b>&ndash;eval-fasta</b></p>
<p class="startli">Assume that alignment for evaluation (cf. &ndash;evaluate) is in fasta format.</p>
</li>
</ul>
<h2><a class="anchor" id="autotoc_md29"></a>
Constraints</h2>
<ul>
<li><p class="startli"><b>&ndash;anchor-constraints=&lt;file</b>&gt;</p>
<p class="startli">Read anchor constraints from bed format specification.</p>
<p class="startli">Anchor constraints in four-column bed format specify positions of named anchor regions per sequence. The 'contig' names have to correspond to the fasta input sequence names. Anchor names must be unique per sequence and regions of the same name for different sequences must have the same length. This constrains the alignment to align all regions of the same name.</p>
<p class="startli">The specification of anchors via this option removes all anchor definitions that may be given directly in the fasta input file!</p>
</li>
<li><p class="startli"><b>&ndash;ignore-constraints</b></p>
<p class="startli">Ignore all constraints (anchor and structure constraints) even if given.</p>
</li>
</ul>
<h2><a class="anchor" id="autotoc_md30"></a>
Rna folding (RNAfold/RNAplfold)</h2>
<ul>
<li><p class="startli"><b>&ndash;noLP</b> / <b>&ndash;LP</b></p>
<p class="startli">Disallow/Allow lonely pairs (default: Disallow).</p>
</li>
<li><p class="startli"><b>&ndash;maxBPspan</b></p>
<p class="startli">Limit maximum span of base pairs (default off).</p>
</li>
<li><p class="startli"><b>&ndash;relaxed-anchors</b></p>
<p class="startli">Relax semantics of anchor constraints (default off, meaning 'strict' semantics). For lexicographically ordered anchors, where each sequence is annotated with exactly the same names, both semantics are equivalent; thus, in this common case, the subtle differences can be ignored. In strict semantics, anchor names must be ordered lexicographically and can only be aligned in this order. In relaxed semantics, the only requirement is that equal anchor names are matched. Consequently, anchor names that don't occur in all sequences could be overwritten (if two names are assigned to the same position) or even introduce inconsistencies.</p>
</li>
<li><p class="startli"><b>&ndash;plfold-span=span</b></p>
<p class="startli">Use RNAplfold with span.</p>
</li>
<li><p class="startli"><b>&ndash;plfold-winsize=ws</b></p>
<p class="startli">Use RNAplfold with window of size ws (default=2*span).</p>
</li>
<li><p class="startli"><b>&ndash;rnafold-parameter=&lt;file</b>&gt;</p>
<p class="startli">Parameter file for RNAfold (RNAfold's -P option)</p>
</li>
<li><p class="startli"><b>&ndash;rnafold-temperature=&lt;temp</b>&gt;</p>
<p class="startli">Temperature for RNAfold (RNAfold's -T option)</p>
</li>
<li><p class="startli"><b>&ndash;skip-pp</b></p>
<p class="startli">Skip computation of pair probs if the probabilities are already existing. Non-existing ones are still computed.</p>
</li>
<li><p class="startli"><b>&ndash;no-bpp-precomputation</b></p>
<p class="startli">Switch off precomputation of base pair probabilties. Overwrite potentially existing input files. (compare skip-pp). For use with special pairwise aligners (e.g. locarna_n) that recompute the base pair probabilities at each invokation.</p>
</li>
<li><p class="startli"><b>&ndash;in-loop-probabilities</b></p>
<p class="startli">Turn on precomputation of in loop probabilties. For use with special pairwise aligners (e.g. locarna_n) that use such probabilities.</p>
</li>
</ul>
<h2><a class="anchor" id="autotoc_md31"></a>
Multithreading</h2>
<ul>
<li><p class="startli"><b>&ndash;threads, &ndash;cpus</b>=number</p>
<p class="startli">Use the given number of threads for computing pair probabilities and all-2-all alignments in parallel (multicore/processor support).</p>
<p class="startli">Be aware: mlocarna seems not to scale well for more than a few threads (often only 2 or 3). Using more threads is often detrimental, since it strongly increases memory consumption due to the current perl threading implementation. This unfortunate behavior seems hard to improve without major rewrite of the software.</p>
</li>
</ul>
<h2><a class="anchor" id="autotoc_md32"></a>
Getting Help</h2>
<ul>
<li><p class="startli"><b>&ndash;help</b></p>
<p class="startli">Brief help message</p>
</li>
<li><p class="startli"><b>&ndash;man</b></p>
<p class="startli">Full documentation</p>
</li>
</ul>
<p>The sequences are given in input file &lt;file&gt; in mfasta format. All results are written to a target directory &lt;dir&gt;. If the file tree is given, contained tree (in NEWICK-tree format) is used as guide tree for the progressive alignment. The final results are collected in &lt;tgtdir&gt;/results. The final multiple alignment is &lt;tgtdir&gt;/results/result.aln.</p>
<h1><a class="anchor" id="autotoc_md33"></a>
EXAMPLES</h1>
<h2><a class="anchor" id="autotoc_md34"></a>
Calling mlocarna</h2>
<p>[Note that the <a class="el" href="namespaceLocARNA.html">LocARNA</a> distribution provides files of the following and other examples in Data/Examples.]</p>
<p>Sequences are typically given in plain fasta format like </p><pre class="fragment">example.fa
----------------------------------------
&gt;fruA
CCUCGAGGGGAACCCGAAAGGGACCCGAGAGG
&gt;fdhA
CGCCACCCUGCGAACCCAAUAUAAAAUAAUACAAGGGAGCAGGUGGCG
&gt;vhuU
AGCUCACAACCGAACCCAUUUGGGAGGUUGUGAGCU
----------------------------------------
</pre><p> To align these sequences, simply call </p><pre class="fragment">mlocarna example.fa
</pre><p> Usually, it makes sense to set additional options; this is either done on the command line or via configuration files. A reasonable small configuration for global alignment of large instances would be </p><pre class="fragment">short-example.cfg
----------------------------------------
max-diff-am: 25
max-diff:    60
min-prob:    0.01
plfold-span: 100
indel:       -50
indel-open:  -750
threads:     8   # &lt;- adapt to your hardware
alifold-consensus-dp
----------------------------------------
</pre><p> To use it, call </p><pre class="fragment">mlocarna --config short-example.cfg example.fa
</pre><p> which is equivalent to </p><pre class="fragment">mlocarna --max-diff-am 25 --max-diff 60 --min-prob 0.01 \
         --indel -50 --indel-open -750 \
         --plfold-span 100 --threads 8 --alifold-consensus-dp \
         example.fa
</pre><p> For probabilistic alignment with consistency transformation, call </p><pre class="fragment">mlocarna --probabilistic --consistency-transform example.fa
</pre><p> In both cases, mlocarna writes the main results to stdout and more detailed results to the target directory example.out. The results directory is overwritten if it exists already. To avoid this, one can specify the target directory (&ndash;tgtdir).</p>
<h2><a class="anchor" id="autotoc_md35"></a>
Use of constraints</h2>
<p>Mlocarna supports structure constraints for folding and anchor constraints for alignment. Both types of constraints can be specified in extension of the standard fasta format via 'constraint lines'. Fasta-ish input with constraints looks like this </p><pre class="fragment">example-w-constraints.fa
----------------------------------------
&gt;A
GACCCUGGGAACAUUAACUACUCUCGUUGGUGAUAAGGAACA
..((.(....xxxxxx...................))).xxx #S
..........000000.......................111 #1
..........123456.......................123 #2
&gt;B
ACGGAGGGAAAGCAAGCCUUCUGCGACA
.(((....xxxxxx.......))).xxx #S
........000000...........111 #1
........123456...........123 #2
----------------------------------------
</pre><p> The same anchor constraints (like by the lines tagged #1, #2) can alternatively be specified in bed format by the entries </p><pre class="fragment">example-anchors.bed
----------------------------------------
A   10      16      first_box
B   8       14      first_box
A   39      42      ACA-box
B   25      28      ACA-box
----------------------------------------
</pre><p> where anchor regions (boxes) have arbitrary but matching names and contig/sequence names correspond to the sequence names of the fasta(-like) input.</p>
<p>Given, e.g. </p><pre class="fragment">example-wo-anchors.fa
----------------------------------------
&gt;A
GACCCUGGGAACAUUAACUACUCUCGUUGGUGAUAAGGAACA
..((.(....xxxxxx...................))).xxx #S
&gt;B
ACGGAGGGAAAGCAAGCCUUCUGCGACA
.(((....xxxxxx.......))).xxx #S
----------------------------------------
</pre><p> once calls </p><pre class="fragment">mlocarna --anchor-constraints example-anchors.bed  example-wo-anchors.fa
</pre> <h2><a class="anchor" id="autotoc_md36"></a>
Realignment</h2>
<p>In realignment mode (option &ndash;realign), mlocarna is called with an input alignment in clustal format, e.g. </p><pre class="fragment">mlocarna --realign example-realign.aln
</pre><p> This allows to define constraints as 'consensus constraints' in the input, e.g. </p><pre class="fragment">example-realign.aln
----------------------------------------
fruA               --CCUCGAGGGGAACCCGAA-------------AGGGACCCGAGAGG--
vhuU               AGCUCACAACCGAACCCAUU-------------UGGGAGGUUGUGAGCU
fdhA               CGCCACCCUGCGAACCCAAUAUAAAAUAAUACAAGGGAGCAG-GUGGCG
#A1                ..*...........CCC.............................5..
#S                 ((((((.((((...(((.................))).)))).))))))
----------------------------------------
</pre><p> Note that anchor names are arbitrary and the consensus structure is 'projected' to the single sequences. Moreover, the input alignment can be used as reference for fast limited realignment, e.g. call to realign in distance 5 of the reference alignment: </p><pre class="fragment">mlocarna --realign example-realign.aln --max-diff 5 --max-diff-aln .
</pre> <h1><a class="anchor" id="autotoc_md37"></a>
AUTHORS</h1>
<p>Sebastian Will Christina Otto (ExpaRNA-P, sparsification classes for ExpaRNA-P and SPARSE) Milad Miladi (SPARSE)</p>
<h1><a class="anchor" id="autotoc_md38"></a>
ONLINE INFORMATION</h1>
<p>For download and online information, see <a href="https://github.com/s-will/LocARNA">https://github.com/s-will/LocARNA</a> and <a href="http://www.bioinf.uni-freiburg.de/Software/LocARNA">http://www.bioinf.uni-freiburg.de/Software/LocARNA</a>.</p>
<p>Latest releases are available as source code on Github at <a href="https://github.com/s-will/LocARNA/releases">https://github.com/s-will/LocARNA/releases</a>.</p>
<h1><a class="anchor" id="autotoc_md39"></a>
REFERENCES</h1>
<p>Sebastian Will, Kristin Reiche, Ivo L. Hofacker, Peter F. Stadler, and Rolf Backofen. Inferring non-coding RNA families and classes by means of genome-scale structure-based clustering. PLOS Computational Biology, 3 no. 4 pp. e65, 2007. doi:10.1371/journal.pcbi.0030065</p>
<p>Sebastian Will, Tejal Joshi, Ivo L. Hofacker, Peter F. Stadler, and Rolf Backofen. LocARNA-P: Accurate boundary prediction and improved detection of structural RNAs. RNA, 18 no. 5 pp. 900-914, 2012. doi:10.1261/rna.029041.111</p>
<p>Sebastian Will, Michael Yu, and Bonnie Berger. Structure-based Whole Genome Realignment Reveals Many Novel Non-coding RNAs. Genome Research, no. 23 pp. 1018-1027, 2013. doi:10.1101/gr.137091.111 </p>
</div></div><!-- contents -->
</div><!-- PageDoc -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Tue Mar 29 2022 16:02:21 for LocARNA-2.0.0RC10 by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.1
</small></address>
</body>
</html>
